File: orchestrator/src/team/team.ts
export enum TeamStatus {
  CREATED = 'CREATED',
  RUNNING = 'RUNNING',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
  DISSOLVED = 'DISSOLVED',
}

export interface TeamRuntime {
  type: 'local' | 'remote';
  workspace: string;
  executorUrl?: string;
}

export interface SkillPack {
  id: string;
  name: string;
  version: string;
  description: string;
  category: string;
  requiredTools: string[];
  systemPrompt: string;
  knowledgeDir?: string;
  rulesDir?: string;
}

export interface TeamMember {
  id: string;
  roleName: string;
  teamRole: 'Coordinator' | 'Planner' | 'Executor' | 'Reviewer' | 'Monitor';
  tools: string[];
  dangerousTools: string[];
  skillPackId?: string;
  status: 'idle' | 'working' | 'speaking' | 'done';
}

export interface Team {
  id: string;
  projectId: string;
  runtime: TeamRuntime;
  members: TeamMember[];
  leader?: TeamMember;
  status: TeamStatus;
}

export interface DagTask {
  taskId: string;
  name: string;
  requiredSkills: string[];
  description?: string;
}

export interface Dag {
  tasks: DagTask[];
}

export function createTeam(id: string, projectId: string, runtime: TeamRuntime): Team {
  return {
    id,
    projectId,
    runtime,
    members: [],
    leader: undefined,
    status: TeamStatus.CREATED,
  };
}

export function addMember(team: Team, member: TeamMember): Team {
  return {
    ...team,
    members: [...team.members, member],
  };
}

export function setLeader(team: Team, memberId: string): Team {
  const member = team.members.find((m) => m.id === memberId);
  if (!member) {
    throw new Error(`Member ${memberId} not found`);
  }
  return {
    ...team,
    leader: member,
  };
}

export function getMemberByTeamRole(
  team: Team,
  teamRole: TeamMember['teamRole'],
): TeamMember | undefined {
  return team.members.find((m) => m.teamRole === teamRole);
}

export function getMemberById(team: Team, id: string): TeamMember | undefined {
  return team.members.find((m) => m.id === id);
}

File: backend/team.py
"""Team — 团队抽象层

管理一组共享运行环境的RoleAgent实例。
Team内部通过Meeting进行讨论和协调。
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field
from typing import Optional


class RuntimeType(str, enum.Enum):
    LOCAL_DOCKER = "local_docker"
    REMOTE_POD = "remote_pod"


class AgentLocation(str, enum.Enum):
    LOCAL = "local"
    REMOTE = "remote"


class TeamStatus(str, enum.Enum):
    CREATED = "created"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    DISSOLVED = "dissolved"


@dataclass
class TeamRuntime:
    """Team共享的运行环境"""
    runtime_id: str
    runtime_type: RuntimeType
    root_path: str
    network_config: dict = field(default_factory=dict)


@dataclass
class TeamMember:
    """Team成员（RoleAgent在Team中的表示）"""
    agent_id: str
    role_name: str
    team_role: str  # Coordinator | Planner | Executor | Reviewer | Monitor
    location: AgentLocation
    skill_pack_id: str = ""
    status: str = "idle"
    member_type: str = "agent"   # "agent" | "human"（human=现实员工，作为把关人）
    approver_for: tuple = ()     # human 成员负责把关的 task_id 列表


@dataclass
class Team:
    """团队实例，管理一组共享Runtime的RoleAgent"""
    team_id: str
    project_id: str
    runtime: TeamRuntime
    members: list[TeamMember] = field(default_factory=list)
    leader: Optional[TeamMember] = field(default=None, repr=False)
    status: TeamStatus = TeamStatus.CREATED

    def add_member(self, member: TeamMember) -> None:
        self.members.append(member)

    def set_leader(self, agent_id: str) -> None:
        for m in self.members:
            if m.agent_id == agent_id:
                self.leader = m
                return
        raise ValueError(f"成员不存在: {agent_id}")

    def set_status(self, status: TeamStatus) -> None:
        self.status = status

    def get_member_by_team_role(self, team_role: str) -> Optional[TeamMember]:
        for m in self.members:
            if m.team_role == team_role:
                return m
        return None

    def get_member_by_id(self, agent_id: str) -> Optional[TeamMember]:
        for m in self.members:
            if m.agent_id == agent_id:
                return m
        return None
